File: image-server/src/main/java/de/haumacher/imageServer/AlbumDate.java
/*
 * Copyright (c) 2026 Bernhard Haumacher et al. All Rights Reserved.
 */
package de.haumacher.imageServer;

import de.haumacher.imageServer.shared.model.AlbumInfo;
import de.haumacher.imageServer.shared.model.AlbumPart;
import de.haumacher.imageServer.shared.model.FolderInfo;
import de.haumacher.imageServer.shared.model.FolderResource;
import de.haumacher.imageServer.shared.model.ImageGroup;
import de.haumacher.imageServer.shared.model.ImagePart;
import de.haumacher.imageServer.shared.model.ListingInfo;
import java.time.Instant;
import java.time.LocalDate;
import java.time.ZoneId;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * When an album happened, and how precisely that is known, see issue #48.
 *
 * <p>
 * There are two dates on the wire and one in the sidecar: {@link AlbumInfo#getDate()} is the
 * explicit date the author set and the only one that is ever stored;
 * {@link AlbumInfo#getEffectiveDate()} is what the album is sorted and placed by and is derived
 * here on every read. This class is the one place that decides what "the date of an album" means.
 * </p>
 *
 * <p>
 * Two resolutions are needed, and they differ by what they are allowed to cost:
 * </p>
 * <ul>
 * <li>{@link #ofAlbum(AlbumInfo, String)} is the full rule — the explicit date, else the leading
 * date of the folder name, else the earliest image date — and is answered to a reader of a single
 * album, whose images the server has loaded anyway.</li>
 * <li>{@link #ofFolder(FolderResource, String)} is the cheap rule — the explicit date, else the
 * folder name — and is what a listing sorts by and what a {@link PlacementRule} files by. Neither
 * of those may open a thousand albums to answer one request.</li>
 * </ul>
 *
 * @author <a href="mailto:haui@haumacher.de">Bernhard Haumacher</a>
 */
public final class AlbumDate {

	/** The separators a date in a folder name may use, as in the album titles of this server. */
	private static final String SEP = "[-_\\.\\s]";

	/**
	 * The date at the start of a folder name: <code>YYYY</code>, <code>YYYY-MM</code> or
	 * <code>YYYY-MM-DD</code>.
	 *
	 * <p>
	 * Anchored at the start and stopped before a further digit, so that <code>20200524 Trip</code>
	 * is not read as the year 2020.
	 * </p>
	 */
	private static final Pattern LEADING_DATE = Pattern.compile(
		"^(\\d{4})(?:" + SEP + "(\\d{2})(?:" + SEP + "(\\d{2}))?)?(?![0-9])");

	/** The date of an album that nothing says anything about. */
	public static final AlbumDate NONE = new AlbumDate(0L, false);

	private final long _millis;

	private final boolean _monthKnown;

	private AlbumDate(long millis, boolean monthKnown) {
		_millis = millis;
		_monthKnown = monthKnown;
	}

	/** The date in milliseconds since the epoch, <code>0</code> if there is none. */
	public long millis() {
		return _millis;
	}

	/** Whether this date says anything at all. */
	public boolean isSet() {
		return _millis > 0L;
	}

	/**
	 * Whether the month of this date is known.
	 *
	 * <p>
	 * A folder named <code>2020 Trip</code> says a year and nothing more; filing it under
	 * <code>2020-01</code> would state a month nobody gave, see
	 * {@link PlacementRule#placementFor(java.nio.file.Path, AlbumDate)}.
	 * </p>
	 */
	public boolean isMonthKnown() {
		return _monthKnown;
	}

	/** The calendar year of this date in the server's time zone; <code>0</code> if it is not set. */
	public int year() {
		return isSet() ? local().getYear() : 0;
	}

	/** The calendar month of this date in the server's time zone; <code>0</code> if it is not set. */
	public int month() {
		return isSet() ? local().getMonthValue() : 0;
	}

	private LocalDate local() {
		return LocalDate.ofInstant(Instant.ofEpochMilli(_millis), ZoneId.systemDefault());
	}

	/**
	 * An explicit date: an instant a person picked, so its month is known.
	 *
	 * @param millis
	 *        Milliseconds since the epoch; <code>0</code> or less means {@link #NONE}.
	 */
	public static AlbumDate ofMillis(long millis) {
		return millis > 0L ? new AlbumDate(millis, true) : NONE;
	}

	/**
	 * The date at the start of the given folder name, {@link #NONE} if it does not start with one.
	 *
	 * <p>
	 * A bare year is January 1st, a year and month the first of that month, both at local midnight;
	 * a year alone is remembered as such, see {@link #isMonthKnown()}. A date that is no date
	 * (<code>2020-13</code>, <code>2020-02-31</code>) falls back to the part of it that is one.
	 * </p>
	 */
	public static AlbumDate ofFolderName(String folderName) {
		if (folderName == null) {
			return NONE;
		}
		Matcher matcher = LEADING_DATE.matcher(folderName);
		if (!matcher.lookingAt()) {
			return NONE;
		}
		int year = Integer.parseInt(matcher.group(1));
		int month = matcher.group(2) == null ? 0 : Integer.parseInt(matcher.group(2));
		int day = matcher.group(3) == null ? 0 : Integer.parseInt(matcher.group(3));

		if (month < 1 || month > 12) {
			return atMidnight(year, 1, 1, false);
		}
		if (day < 1 || day > LocalDate.of(year, month, 1).lengthOfMonth()) {
			return atMidnight(year, month, 1, true);
		}
		return atMidnight(year, month, day, true);
	}

	/**
	 * The date a listing sorts a folder by and a {@link PlacementRule} files it by: the explicit
	 * date of its sidecar, else the date in its name.
	 *
	 * <p>
	 * No image is opened here, and none is even looked at: this is the date that can be answered
	 * for every folder of a listing without loading the albums behind them.
	 * </p>
	 *
	 * @param sidecar
	 *        The folder's own <code>index.json</code>, <code>null</code> if it has none.
	 * @param folderName
	 *        The name of the folder on disk.
	 */
	public static AlbumDate ofFolder(FolderResource sidecar, String folderName) {
		if (sidecar instanceof AlbumInfo) {
			AlbumDate explicit = ofMillis(((AlbumInfo) sidecar).getDate());
			if (explicit.isSet()) {
				return explicit;
			}
		}
		return ofFolderName(folderName);
	}

	/**
	 * The date the given album is sorted and placed by: the explicit date, else the date in the
	 * folder name, else the earliest date of an image it holds.
	 *
	 * @param album
	 *        The album as it was loaded, images included.
	 * @param folderName
	 *        The name of the album's folder on disk.
	 */
	public static AlbumDate ofAlbum(AlbumInfo album, String folderName) {
		AlbumDate byFolder = ofFolder(album, folderName);
		if (byFolder.isSet()) {
			return byFolder;
		}
		return ofMillis(earliestImageDate(album));
	}

	/**
	 * The earliest {@link ImagePart#getDate() date} of the images of the given album, <code>0</code>
	 * if it holds none with a date.
	 */
	public static long earliestImageDate(AlbumInfo album) {
		long result = 0L;
		for (AlbumPart part : album.getParts()) {
			if (part instanceof ImagePart) {
				result = earlier(result, ((ImagePart) part).getDate());
			} else if (part instanceof ImageGroup) {
				for (ImagePart image : ((ImageGroup) part).getImages()) {
					result = earlier(result, image.getDate());
				}
			}
		}
		return result;
	}

	/**
	 * Clears everything the server derives from a folder resource before it is written to disk.
	 *
	 * <p>
	 * The sidecar doctrine: what is stored means what it says, and
	 * {@link AlbumInfo#getEffectiveDate()} says "this is what the album is sorted by today", which
	 * is not a statement anybody made. Every path that writes an <code>index.json</code> goes
	 * through here, so a derived date can never be frozen into a sidecar by a round trip, see
	 * {@link ImageServlet#storeSidecar(java.io.File, byte[])}.
	 * </p>
	 *
	 * @return Whether anything had to be cleared.
	 */
	public static boolean clearDerived(FolderResource resource) {
		boolean changed = false;
		if (resource instanceof AlbumInfo) {
			AlbumInfo album = (AlbumInfo) resource;
			if (album.getEffectiveDate() != 0L) {
				album.setEffectiveDate(0L);
				changed = true;
			}
		} else if (resource instanceof ListingInfo) {
			// The folders of a stored listing are rebuilt from the disk on every read; a derived
			// date among them would be stale the moment it was written.
			for (FolderInfo folder : ((ListingInfo) resource).getFolders()) {
				if (folder.getEffectiveDate() != 0L) {
					folder.setEffectiveDate(0L);
					changed = true;
				}
			}
		}
		return changed;
	}

	private static long earlier(long result, long candidate) {
		if (candidate <= 0L) {
			return result;
		}
		return result <= 0L ? candidate : Math.min(result, candidate);
	}

	private static AlbumDate atMidnight(int year, int month, int day, boolean monthKnown) {
		long millis = LocalDate.of(year, month, day).atStartOfDay(ZoneId.systemDefault()).toInstant().toEpochMilli();
		return new AlbumDate(millis, monthKnown);
	}

	@Override
	public String toString() {
		return isSet() ? local().toString() + (_monthKnown ? "" : " (year only)") : "<no date>";
	}
}

File: valbum_ui/lib/album_date.dart
/// When an album happened: the date the app shows, where that date comes
/// from, and the order a listing is read in, see issue #48.
///
/// This library is pure — it knows nothing about widgets or the transport —
/// so the two rules it holds can be unit-tested on their own:
///
/// * [describeDateSource] answers where the date on the screen comes from, so
///   the album properties can say it instead of showing a date out of nowhere.
/// * [sortedFolders] is the order the server's `BY_DATE` comparator produces
///   (`ResourceCache.BY_DATE`): the newest first, the undated behind them by
///   name. The app applies it to what it was given, so a listing from an older
///   server or from the offline cache reads like a current one.
///
/// Two dates travel on the wire and only one is ever stored: [AlbumInfo.date]
/// is the explicit date the author set, [AlbumInfo.effectiveDate] and
/// [FolderInfo.effectiveDate] are derived by the server on every read. The app
/// therefore never *sets* an effective date; it only reads one.
library;

import 'package:intl/intl.dart';

import 'resource.dart';

/// Where the date an album is shown and sorted under comes from.
enum DateSource {
  /// Nothing says when the album happened.
  none,

  /// The author set the date, see [AlbumInfo.date].
  explicit,

  /// The album's folder name begins with the date.
  folderName,

  /// The date of the earliest photo of the album.
  photos,
}

/// Where the effective date of [album] comes from.
///
/// The server derives [AlbumInfo.effectiveDate] in this order: the explicit
/// [AlbumInfo.date], else the leading date of the album's folder name, else
/// the earliest image date (see `AlbumDate.ofAlbum`). The app cannot see the
/// image dates of a folder it has not opened, but it can see the two cheap
/// sources — and what is neither of them is the third.
///
/// [folderName] is the name of the folder the album lives in. The server does
/// not fill [Resource.path] of a loaded album, so the view passes the last
/// segment of the path it is showing; the album's own path is the fallback.
DateSource describeDateSource(AlbumInfo album, {String? folderName}) {
  if (album.date != 0) {
    return DateSource.explicit;
  }
  if (album.effectiveDate == 0) {
    return DateSource.none;
  }
  var fromName = folderNameDate(folderName ?? folderNameOf(album.path));
  if (fromName != null &&
      fromName.millisecondsSinceEpoch == album.effectiveDate) {
    return DateSource.folderName;
  }
  return DateSource.photos;
}

/// The name of the folder a resource path addresses, the last segment.
///
/// A resource path may name the album with more than one segment (or with
/// none, at the root); the date is written on the folder itself.
String folderNameOf(String path) {
  var segments = path.split("/").where((part) => part.isNotEmpty).toList();
  return segments.isEmpty ? "" : segments.last;
}

/// The separators a date in a folder name may use, as `AlbumDate.SEP` does.
const String _sep = r"[-_.\s]";

/// The date at the start of a folder name: `YYYY`, `YYYY-MM` or
/// `YYYY-MM-DD`, anchored at the start and stopped before a further digit, so
/// that `20200524 Trip` is not read as the year 2020.
final RegExp _leadingDate =
    RegExp("^(\\d{4})(?:$_sep(\\d{2})(?:$_sep(\\d{2}))?)?(?![0-9])");

/// The date at the start of [folderName], `null` if it does not start with
/// one.
///
/// A bare year is January 1st, a year and month the first of that month, both
/// at local midnight — exactly what the server's `AlbumDate.ofFolderName`
/// answers, so that a date read here compares with the effective date the
/// server derived from the same name. A date that is no date (`2020-13`,
/// `2020-02-31`) falls back to the part of it that is one.
DateTime? folderNameDate(String folderName) {
  var match = _leadingDate.matchAsPrefix(folderName);
  if (match == null) {
    return null;
  }
  var year = int.parse(match.group(1)!);
  var month = match.group(2) == null ? 0 : int.parse(match.group(2)!);
  var day = match.group(3) == null ? 0 : int.parse(match.group(3)!);

  if (month < 1 || month > 12) {
    return DateTime(year);
  }
  if (day < 1 || day > _daysInMonth(year, month)) {
    return DateTime(year, month);
  }
  return DateTime(year, month, day);
}

int _daysInMonth(int year, int month) =>
    DateTime(year, month + 1).difference(DateTime(year, month)).inDays;

/// The folders of a listing in the order they are shown: the newest first,
/// the undated behind them by name.
///
/// A copy, never the list it was given: the listing the app holds is what the
/// server said, and the order is how it is read out.
List<FolderInfo> sortedFolders(List<FolderInfo> folders) {
  var result = List.of(folders);
  result.sort((a, b) {
    if (a.effectiveDate != b.effectiveDate) {
      // Newest first; an undated folder (0) therefore falls to the end.
      return b.effectiveDate.compareTo(a.effectiveDate);
    }
    return a.name.toLowerCase().compareTo(b.name.toLowerCase());
  });
  return result;
}

/// The date of an album as the app writes it out for a reader, `null` when
/// nothing says when the album happened, see issue #107.
///
/// [effectiveDate] is what the server derived (`AlbumInfo.effectiveDate`,
/// `FolderInfo.effectiveDate`), `0` for none. The form is the locale's own
/// medium date, so the listing tile and the album's own heading say the date
/// the way the rest of the app does — and never the folder-name spelling,
/// which is a naming convention and not a thing to read.
String? albumDateLabel(int effectiveDate) => effectiveDate == 0
    ? null
    : DateFormat.yMMMd()
        .format(DateTime.fromMillisecondsSinceEpoch(effectiveDate));

/// Whether a listing entry is one a date is shown for, see issue #133.
///
/// Only an album happened on a day. A folder of folders carries an
/// [FolderInfo.effectiveDate] all the same — it is what the listing is sorted
/// by, and a folder named `2026` sorts with the year it names — but that is a
/// sort key, not a date: showing it read "2026 - Jan 1 2026".
///
/// [FolderKind.album] is the value an entry from a server that does not know
/// the field yet carries, so such a listing shows its dates exactly as before.
bool folderHasDate(FolderInfo folder) =>
    folder.kind == FolderKind.album && folder.effectiveDate != 0;

/// The spelling of a date in a folder name, the naming convention `yyyy-MM-dd`.
final DateFormat folderDateFormat = DateFormat("yyyy-MM-dd");

/// The folder name an album with [title] taken on [date] is created under —
/// and, since issue #130, the name its folder is renamed to whenever its
/// properties are written.
///
/// `yyyy-MM-dd title` by the naming convention, and the title alone when
/// [date] is `null` — an album without a date is no less an album, it is only
/// one the placement rules leave where it was made, see issue #119. The title
/// is trimmed: a stray blank at either end is a slip of the keyboard, never a
/// name.
///
/// The server composes the very same name (`FolderNames.albumFolderName`), and
/// the two are pinned against each other by the shared table
/// `image-server/src/test/fixtures/folder-names.json`, so a rule changed on
/// one side and not on the other fails the tests of both.
String albumFolderName(DateTime? date, String title) {
  var name = title.trim();
  if (date == null) {
    return name;
  }
  var day = folderDateFormat.format(date);
  return name.isEmpty ? day : "$day $name";
}

/// The name of the folder a folder of folders with [title] carries, see
/// issue #130.
///
/// Its title alone: a folder of folders has no date of its own, only the
/// dates of what lies in it.
String listingFolderName(String title) => title.trim();

/// Whether [name] may be the name of a folder on disk.
///
/// A name that is a path (`a/b`), a name that is a navigation step (`.`,
/// `..`) and a name that hides the folder (`.thing`, which is what the server
/// itself uses for `.valbum` and `.vacache`) are none. The twin of the
/// server's `FolderNames.isLegal`, pinned by the same shared table: the app
/// says so before the round trip, and the server refuses it with a `400`
/// either way.
bool isLegalFolderName(String name) {
  if (name.trim().isEmpty) {
    return false;
  }
  if (name == "." || name == ".." || name.startsWith(".")) {
    return false;
  }
  return !name.contains("/") && !name.contains(r"\") && !name.contains("\u0000");
}
